### PR TITLE
Refactor project layout for better go build and get ergonomics

### LINKED ISSUE #7: Pull request body validation error
```
DEBU[0010] Successfully pushed local branch to remote origin  Repo=terraform-aws-architecture-catalog
DEBU[0011] Error opening Pull request                    Base=master Body="_**This PR was created by git-xargs. See https://github.com/gruntwork-io/prototypes/pull/152 for context.**_ We recently renamed all of our repos to follow the Terraform registry format of  (e.g., ). This PR does a search & replace to update all references to these repos to the new names. GitHub can handle redirects, so in theory, everything should work fine with the old names, but there were so many renames, that to reduce confusion, this will update most of our references to the proper values." Error="POST https://api.github.com/repos/gruntwork-io/terraform-aws-architecture-catalog/pulls: 422 Validation Failed [{Resource:PullRequest Field:base Code:invalid Message:}]" Head=refs/heads/rename-repos
DEBU[0011] Error encountered while processing repo       Error="POST https://api.github.com/repos/gruntwork-io/terraform-aws-architecture-catalog/pulls: 422 Validation Failed [{Resource:PullRequest Field:base Code:invalid Message:}]" Repo name=terraform-aws-architecture-catalog

```

### PR DESCRIPTION
**Pull request by zackproser**
Fixes https://github.com/gruntwork-io/git-xargs/issues/18
Fixes #7 

@mmlb  Please have a look and let me know if this is what you had in mind. Thanks!

Updates project layout which will allow for installing `git-xargs` with a single call to `go get github.com/gruntwork-io/git-xargs`.

### CODE REVIEW

**Review by yorinasub17 — COMMENTED** [review]
Do you also need to update https://github.com/gruntwork-io/git-xargs/blob/master/.circleci/config.yml?

**Comment by zackproser** [comment]
> Do you also need to update https://github.com/gruntwork-io/git-xargs/blob/master/.circleci/config.yml?

Yes, thank you! Done in 8af5012

File: README.md [hunk 1]
```diff
@@ -153,6 +155,18 @@ COMMAND SUPPLIED
     git-xargs --version
     ```
 
+### Installion option 2: Run go get 
+
+1. **Run go get to install the latest release of git-xargs**
+
+`go get github.com/gruntwork-io/git-xargs`
```
Comment (mmlb) [inline README.md:162]:
```suggestion
`go get github.com/gruntwork-io/git-xargs/cmd/git-xargs`
```

File: README.md [hunk 2]
```diff
@@ -153,6 +155,18 @@ COMMAND SUPPLIED
...
+
+1. **Alternatively, use go get to install a specific release of git-xargs**
+
+`go get github.com/gruntwork-io/git-xargs@v0.0.4`
```
Comment (mmlb) [inline README.md:166]:
```suggestion
`go get github.com/gruntwork-io/git-xargs/cmd/git-xargs@v0.0.4`
```
I'm not actually sure if this one works.

**Review by mmlb — CHANGES_REQUESTED** [review]

**Comment by mmlb** [comment]
> Fixes #18
> 
> @mmlb Please have a look and let me know if this is what you had in mind. Thanks!

Yep, thanks.

**Review by mmlb — DISMISSED** [review]
lgtm

**Comment by sc250024** [comment]
This is needed in addition to the small PR that I just merged: https://github.com/gruntwork-io/git-xargs/pull/14. I realized that after it was merged.

If you now run `go get -u -v github.com/gruntwork-io/git-xargs/cmd` against https://github.com/gruntwork-io/git-xargs/releases/tag/v0.0.4, it "works", but the binary is now `cmd` :D

Example:

```console
$ cmd --help                                                                                                

Usage: cmd [--loglevel] [--github-org] [--dry-run] [--skip-pull-requests] [--repo] [--repos] [--branch-name]
[--commit-message] [--pull-request-title] [--pull-request-description] [--help] command [options] [args]

git-xargs is a command-line tool (CLI) for making updates across multiple Github repositories with a single command.

Commands:

   help, h  Shows a list of commands or help for one command

$ which cmd

/Users/<my-user>/go/bin/cmd
```

File: README.md [hunk 3]
```diff
@@ -153,6 +155,18 @@ COMMAND SUPPLIED
...
 
+### Installion option 2: Run go get 
+
+1. **Run go get to install the latest release of git-xargs**
+
+`go get github.com/gruntwork-io/git-xargs/cmd/git-xargs`
+
+1. **Alternatively, use go get to install a specific release of git-xargs**
+
+`go get github.com/gruntwork-io/git-xargs/cmd/git-xargs@v0.0.4`
```
Comment (sc250024) [inline README.md:166]: I think this should be quickly updated to `v0.0.5` since `v0.0.4` was just released, and doesn't contain this fix.

**Review by brikis98 — COMMENTED** [review]
Why is the folder called `cmd` in the first place?

In our other CLI tools, we typically put `main.go` in the root of the repo, and the packages it depends on within folders. There might be a package called, say, `cli` or `cmd` that contains the core logic for parsing CLI args and executing code, a `git` package for the Git functionality, and so on.

If we put `main.go` in the root, would calling `go get github.com/gruntwork-io/git-xargs` do what we want?

**Comment by sc250024** [comment]
> Why is the folder called `cmd` in the first place?
> 
> In our other CLI tools, we typically put `main.go` in the root of the repo, and the packages it depends on within folders. There might be a package called, say, `cli` or `cmd` that contains the core logic for parsing CLI args and executing code, a `git` package for the Git functionality, and so on.
> 
> If we put `main.go` in the root, would calling `go get github.com/gruntwork-io/git-xargs` do what we want?

@brikis98 I've seen both patterns. Some repos do what you mention, and some put it in `cmd/<name-of-binary>`. Here's an example: https://github.com/jsonnet-bundler/jsonnet-bundler.

Both work, it's just a matter of preference.

**Comment by zackproser** [comment]
> Why is the folder called `cmd` in the first place?
> 
> In our other CLI tools, we typically put `main.go` in the root of the repo, and the packages it depends on within folders. There might be a package called, say, `cli` or `cmd` that contains the core logic for parsing CLI args and executing code, a `git` package for the Git functionality, and so on.
> 
> If we put `main.go` in the root, would calling `go get github.com/gruntwork-io/git-xargs` do what we want?

@brikis98 Raised this discussion internally yesterday - the `cmd` folder originally came from this layout: https://github.com/golang-standards/project-layout#cmd and then we carried that to certain projects including kubergrunt. 

That said, it's not required, and personally I'd prefer to drop the `cmd` path entirely - because the other reason this is tripping some folks up is that they're used to being able to check out a repo and run `go build` in the root, which you also can't currently do. 

I agree calling `go get github.com/gruntwork-io/git-xargs` would be the ideal.

**Comment by mmlb** [comment]
Dropping the cmd/ subfolder entirely works for me too, I was mostly interested in not having a binary named `cmd` in my GOPATH :smile:

**Comment by zackproser** [comment]
Rearranged everything so that: 
* `go build` works from the root directory as expected
* Go source code is split into reasonable packages at the first level of the repo 
* `go test -v ./...` passes 
* This should support the use case of being able to simple `go get github.com/gruntwork-io/git-xargs` 

@brikis98 @yorinasub17 please have a look when you have time

File: README.md (hunk 3 above)
Comment (zackproser) [reply README.md]: Done in e9c1276

**Review by sc250024 — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -137,6 +137,8 @@ COMMAND SUPPLIED
     export GITHUB_OAUTH_TOKEN=<your-secret-github-oauth-token>
     ```
 
+### Installation option 1: Installing published binaries
```
Comment (brikis98) [inline README.md:140]: If you're going to have sub-headings in the getting started instructions, you need to group things more logically. Prob the minimal change here is to take the "Export a valid Github token" step above and move it below, making it the first step in the "Try it out" section.

File: README.md [hunk 5]
```diff
@@ -153,6 +155,18 @@ COMMAND SUPPLIED
     git-xargs --version
     ```
 
+### Installion option 2: Run go get 
+
+1. **Run go get to install the latest release of git-xargs**
```
Comment (brikis98) [inline README.md:160]: This should have as a requisite that you have Go coding env up and running.

File: README.md [hunk 6]
```diff
@@ -366,6 +384,14 @@ Once you've done this, please feel free to open a pull request adding your scrip
 
 Thanks for contributing back! Our hope is that eventually this repo will contain many useful generic scripts for common maintenance and upgrading tasks that everyone can leverage to save time.
 
+## Building the binary from source
+
+Clone this repository and then run the following command from the root of the repository: 
+
+`go build -o git-xargs ./cmd/git-xargs`
```
Comment (brikis98) [inline README.md:391]: `go build` should be enough here. `cmd` doesn't exist anyway. Also, use code fences (three backticks) for code blocks.

File: README.md [hunk 7]
```diff
@@ -385,7 +411,7 @@ This is especially helpful if you are developing against the tool and want to qu
 Tests are included under the `/cmd` directory. There is a mixture of unit tests as well as CLI interface tests that verify certain flags are required, error messages are returned as expected when invalid input is provided, etc.
 
 ```
-go test -v ./cmd
+go test -v ./cmd/git-xargs
```
Comment (brikis98) [inline README.md:414]: There's no such folder now... You probably want to use the triple dot to run tests in all packages. Oh, and if tests take > 10 min, add `-timeout XXX`.

File: README.md (hunk 4 above)
Comment (zackproser) [reply README.md:133]: Done in a16033f. Sorry - the markdown file also got reformatted by my new vim plugin as a result (which I know you love), but I think it was due for, as the table was looking gross. So here's the direct link to the change: https://github.com/gruntwork-io/git-xargs/pull/20/commits/a16033fca4c1218887420748a3a3bb624f85a57a#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R163

File: README.md (hunk 5 above)
Comment (zackproser) [reply README.md:155]: Added in b983994

File: README.md (hunk 6 above)
Comment (zackproser) [reply README.md]: :+1:  0689413

File: README.md (hunk 7 above)
Comment (zackproser) [reply README.md:423]: Roger - thanks for catching that. Tests luckily take ~46 seconds right now so I will omit timeout for the time being.

Comment (zackproser) [reply README.md]: Fixed in 145d9a4 and removed some unnecessary / out of date text.

**Comment by zackproser** [comment]
@brikis98 Feedback implemented. Thank you!

**Comment by zackproser** [comment]
Need to dig into test failures.

**Comment by brikis98** [comment]
Don't forget to pull in the latest from `master` and merging it, since all the file paths have changed!

**Comment by zackproser** [comment]
> Don't forget to pull in the latest from `master` and merging it, since all the file paths have changed!

Did thanks - rebased and now have it all passing again @brikis98 :pray:

**Comment by zackproser** [comment]
Thanks for the reviews! Going to merge this in now.

**Comment by yermulnik** [comment]
Was looking forward for this to be merged so that `go get …` can be used to install binary, but it seems like that it hasn't been released yet so that `go get …` doesn't do the thing yet — what is the ETA for this feature to be released please?

**Comment by yermulnik** [comment]
Oh, nevermind me — have just seen `v0.0.5` was released :partying_face:

**Pull request merged**